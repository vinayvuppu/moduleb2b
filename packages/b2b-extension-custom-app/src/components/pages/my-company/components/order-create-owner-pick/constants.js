// eslint-disable-next-line
export const OWNER_TYPES = {
  OWNER_ME: 'me',
  OWNER_EMPLOYEE: 'employee',
  OWNER_COMPANY: 'company',
};
