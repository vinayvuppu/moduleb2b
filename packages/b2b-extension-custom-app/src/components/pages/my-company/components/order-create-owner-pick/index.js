export { default } from './order-create-owner-pick';
export * from './constants';
