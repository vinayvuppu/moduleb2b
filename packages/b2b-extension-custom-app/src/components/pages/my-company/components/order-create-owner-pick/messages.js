import { defineMessages } from 'react-intl';

export default defineMessages({
  subTitle: {
    id: 'Orders.Create.Step.Owner.subTitle',
    description: 'A subtitle for the customer step',
    defaultMessage: 'Create order for',
  },
  columnCustomerNumber: {
    id: 'Orders.Create.Step.Owner.table.customerNumber',
    description: 'Title of the table column (customerNumber)',
    defaultMessage: 'Employee number',
  },
  columnFirstName: {
    id: 'Orders.Create.Step.Owner.table.firstName',
    description: 'Title of the table column (firstName)',
    defaultMessage: 'First name',
  },
  columnLastName: {
    id: 'Orders.Create.Step.Owner.table.lastName',
    description: 'Title of the table column (lastName)',
    defaultMessage: 'Last name',
  },
  columnCompanyName: {
    id: 'Orders.Create.Step.Owner.table.companyName',
    description: 'Title of the table column (companyName)',
    defaultMessage: 'Company',
  },
  columnEmail: {
    id: 'Orders.Create.Step.Owner.table.email',
    description: 'Title of the table column (email)',
    defaultMessage: 'Email',
  },
  noCustomersText: {
    id: 'Orders.Create.Step.Owner.table.noCustomersText',
    description: 'No customers',
    defaultMessage:
      'There are no Employees that matched with the given input text',
  },
  searchPlaceholder: {
    id: 'Orders.Create.Step.Owner.searchPlaceholder',
    description: 'A searchPlaceholder for the customer search',
    defaultMessage: 'Search by Employee number or Email',
  },
  noSearchResults: {
    id: 'Orders.Create.Step.Owner.noSearchResults',
    description: 'A text when the search results are zero',
    defaultMessage: 'There are no results matching your search criteria.',
  },
  ownerMe: {
    id: 'Orders.Create.Step.Owner.ownerMe',
    defaultMessage: 'Me',
  },
  ownerEmployee: {
    id: 'Orders.Create.Step.Owner.ownerEmployee',
    defaultMessage: 'Company employee',
  },
  ownerCompany: {
    id: 'Orders.Create.Step.Owner.ownerCompany',
    defaultMessage: 'Company',
  },
  ownerSelectButton: {
    id: 'Orders.Create.Step.Owner.ownerSelectButton',
    defaultMessage: 'Select',
  },
  userWithoutStore: {
    id: 'Orders.Create.Step.Owner.userWithoutStore',
    defaultMessage: 'User without store',
  },
});
