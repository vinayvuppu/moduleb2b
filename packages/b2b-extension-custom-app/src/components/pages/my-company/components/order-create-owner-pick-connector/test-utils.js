/* eslint-disable import/prefer-default-export */
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import flatMap from 'lodash.flatmap';
import intersection from 'lodash.intersection';
import FetchCustomersQuery from './order-create-customer-pick-connector.graphql';

export const createCustomersFetchQueryMock = (stores = []) => {
  const whereClause =
    stores.length === 0
      ? undefined
      : `(stores(key = "${stores.join(',')}") or stores is empty)`;
  const customers = [
    {
      id: 'customer-1',
      version: 4,
      firstName: 'Anna',
      lastName: 'Banana',
      email: 'annabanana@blablub.com',
      customerNumber: '111',
      companyName: null,
      customerGroup: null,
      stores: [
        {
          id: 'store-2',
          key: 'usa',
          nameAllLocales: [{ locale: 'en-GB', value: 'USA' }],
        },
      ],
      addresses: [
        {
          id: '_ZqTe0es',
          firstName: '3333',
          streetName: 'Fakestreet',
          streetNumber: '23',
          country: 'DE',
          additionalAddressInfo: 'Additional address info2',
          additionalStreetInfo: 'Additional street info',
          email: null,
          phone: null,
          city: null,
          company: null,
          lastName: null,
          postalCode: null,
          region: null,
        },
      ],
    },
    {
      id: 'customer-2',
      version: 1,
      firstName: 'Ali',
      lastName: 'Baba',
      customerNumber: '235',
      email: 'alibaba@testest.de',
      companyName: null,
      customerGroup: null,
      stores: [
        {
          id: 'store-1',
          key: 'germany',
          nameAllLocales: [{ locale: 'en-GB', value: 'Germany' }],
        },
      ],
      addresses: [],
    },
    {
      id: 'customer-3',
      version: 3,
      customerNumber: '128',
      firstName: 'Amin',
      lastName: 'Rocks',
      email: 'email@ct.de',
      companyName: null,
      customerGroup: null,
      stores: [],
      addresses: [
        {
          id: 'yUQGmqRG',
          firstName: 'first',
          lastName: 'last2',
          company: 'company',
          streetName: 'Street name',
          streetNumber: 'House number',
          city: 'city',
          postalCode: '66215',
          region: 'region',
          country: 'US',
          additionalAddressInfo: 'Additional address info2',
          additionalStreetInfo: 'Additional street info',
          email: 'email2@email.com',
          phone: 'phone',
        },
      ],
    },
  ].filter(customer => {
    // if we don't pass stores then we don't filter
    if (!stores) return true;

    const customerStoresArray = flatMap(customer.stores, s => s.key);
    return (
      // true if the customer belongs to one of the stores passed
      intersection(customerStoresArray, stores).length !== 0 ||
      // or the customer belongs to no store
      customerStoresArray.length === 0
    );
  });

  return {
    request: {
      variables: {
        target: GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM,
        projectKey: 'context-playground',
        limit: 20,
        locale: 'en',
        offset: 0,
        sort: ['createdAt desc'],
        where: whereClause,
      },
      query: FetchCustomersQuery,
    },
    result: {
      data: {
        customers: {
          count: 3,
          offset: 0,
          total: 3,
          results: customers,
        },
      },
    },
  };
};
