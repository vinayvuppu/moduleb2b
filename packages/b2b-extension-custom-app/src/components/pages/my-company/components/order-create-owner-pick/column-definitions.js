import { defaultMemoize } from 'reselect';
import messages from './messages';

export default defaultMemoize(intl => [
  {
    key: 'employeeNumber',
    label: intl.formatMessage(messages.columnCustomerNumber),
    flexGrow: 1,
  },
  {
    key: 'firstName',
    label: intl.formatMessage(messages.columnFirstName),
    flexGrow: 1,
  },
  {
    key: 'lastName',
    label: intl.formatMessage(messages.columnLastName),
    flexGrow: 1,
  },
  {
    key: 'email',
    label: intl.formatMessage(messages.columnEmail),
    flexGrow: 1,
  },
]);
