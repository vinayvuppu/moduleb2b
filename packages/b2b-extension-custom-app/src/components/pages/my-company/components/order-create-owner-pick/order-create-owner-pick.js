import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import { injectIntl, FormattedMessage } from 'react-intl';

import {
  Spacings,
  Table,
  Text,
  LoadingSpinner,
  Card,
  PrimaryButton,
  CheckBoldIcon,
} from '@commercetools-frontend/ui-kit';
import { NO_VALUE_FALLBACK, DOMAINS } from '@commercetools-frontend/constants';
import * as globalActions from '@commercetools-frontend/actions-global';
import SearchInput from '@commercetools-local/core/components/fields/search-input';
import Pagination from '@commercetools-local/core/components/search/pagination';
import PageBottomSpacer from '@commercetools-local/core/components/page-bottom-spacer';
import { selectDefaultAddress } from '../../utils/address-selection';
import EmployeeDetailWrapper from '../employee-detail-wrapper';
import replaceLocationWhenDraftIsMissing from '../../utils/replace-location-when-draft-is-missing';

import messages from './messages';
import { OWNER_TYPES } from './constants';

import OrderCreateOwnerPickConnector from '../order-create-owner-pick-connector';
import createColumnsDefinition from './column-definitions';
import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';

export const ORDER_SET_CUSTOMER_TAB_NAMES = {
  ADDRESSES: 'addresses',
};

export class OrderCreateOwnerPick extends React.PureComponent {
  static displayName = 'OrderCreateOwnerPick';

  static propTypes = {
    title: PropTypes.string.isRequired,
    goToAddressSelection: PropTypes.func.isRequired,
    projectKey: PropTypes.string.isRequired,
    employeesFetcher: PropTypes.shape({
      isLoading: PropTypes.bool.isRequired,
      searchQuery: PropTypes.shape({
        set: PropTypes.func.isRequired,
        value: PropTypes.shape({
          perPage: PropTypes.number.isRequired,
          page: PropTypes.number.isRequired,
          searchText: PropTypes.string,
        }),
      }),
      employees: PropTypes.shape({
        results: PropTypes.array.isRequired,
        total: PropTypes.number.isRequired,
      }),
    }),
    cartDraftState: PropTypes.shape({
      value: PropTypes.shape({
        id: PropTypes.string.isRequired,
        currency: PropTypes.string.isRequired,
      }),
      update: PropTypes.func.isRequired,
    }),
    storeState: PropTypes.shape({
      value: PropTypes.shape({
        id: PropTypes.string.isRequired,
        key: PropTypes.string.isRequired,
        nameAllLocales: PropTypes.arrayOf(
          PropTypes.shape({
            locale: PropTypes.string.isRequired,
            value: PropTypes.string.isRequired,
          })
        ).isRequired,
      }),
      update: PropTypes.func.isRequired,
    }),
    initialSelectionModalState: PropTypes.shape({
      isOpen: PropTypes.bool.isRequired,
      close: PropTypes.func.isRequired,
    }),
    ownerState: PropTypes.shape({
      owner: PropTypes.shape({
        type: PropTypes.string,
        company: PropTypes.object,
      }).isRequired,
      update: PropTypes.func.isRequired,
    }),
    company: PropTypes.object.isRequired,
    employee: PropTypes.object.isRequired,

    // HoC
    intl: PropTypes.shape({ formatMessage: PropTypes.func.isRequired })
      .isRequired,

    // Connected
    showNotification: PropTypes.func.isRequired,
  };

  // This function sets by default the shipping and billing addresses of the
  // cartDraft depending on the default properties. In case the there are no
  // default addresses, the first address will be the selected one for the cartDraft
  changeCustomerDraft = (customer, owner) => {
    if (customer.stores.length) {
      this.props.cartDraftState.update({
        customerId: customer.id,
        customerEmail: customer.email,
        ...selectDefaultAddress(owner),
        store: {
          key: customer.stores[0].key,
          typeId: 'store',
        },
      });
    } else {
      // TODO
      this.props.showNotification({
        kind: 'error',
        domain: DOMAINS.GLOBAL,
        text: this.props.intl.formatMessage(messages.userWithoutStore),
      });
    }
  };

  handleSearchInputChange = searchText =>
    this.props.employeesFetcher.searchQuery.set({
      ...this.props.employeesFetcher.searchQuery.value,
      searchText,
    });

  handlePerPageChange = perPage =>
    this.props.employeesFetcher.searchQuery.set({
      ...this.props.employeesFetcher.searchQuery.value,
      perPage,
    });

  handlePageChange = page =>
    this.props.employeesFetcher.searchQuery.set({
      ...this.props.employeesFetcher.searchQuery.value,
      page,
    });

  renderItem = ({ rowIndex, columnKey }) => {
    const value = this.props.employeesFetcher.employees.results[rowIndex][
      columnKey
    ];

    return value || NO_VALUE_FALLBACK;
  };

  handleRowClick = rowIndex => {
    this.handleSelectOwner({
      type: OWNER_TYPES.OWNER_EMPLOYEE,
      company: undefined,
      employeeId: this.props.employeesFetcher.employees.results[rowIndex].id,
    });

    this.changeCustomerDraft(
      this.props.employee,
      this.props.employeesFetcher.employees.results[rowIndex]
    );
    this.props.goToAddressSelection(
      this.props.employeesFetcher.employees.results[rowIndex].id
    );
  };

  handleSelectOwner = owner => {
    this.props.ownerState.update(owner);
  };

  handleSelectOwnerMe = () => {
    this.handleSelectOwner({
      type: OWNER_TYPES.OWNER_ME,
      company: undefined,
      employeeId: this.props.employee.id,
    });

    this.changeCustomerDraft(this.props.employee, this.props.employee);
    this.props.goToAddressSelection(this.props.employee.id);
  };

  handleSelectOwnerEmployee = () => {
    this.handleSelectOwner({
      type: OWNER_TYPES.OWNER_EMPLOYEE,
      company: undefined,
      employeeId: undefined,
    });
  };
  handleSelectOwnerCompany = () => {
    this.handleSelectOwner({
      type: OWNER_TYPES.OWNER_COMPANY,
      company: this.props.company,
      employeeId: undefined,
    });

    this.changeCustomerDraft(this.props.employee, this.props.company);

    this.props.goToAddressSelection();
  };

  render() {
    if (this.props.employeesFetcher.isLoading) return <LoadingSpinner />;
    return (
      <Fragment>
        {this.props.ownerState.owner.type === undefined && (
          <Spacings.Inline>
            <Card>
              <Spacings.Inline
                justifyContent="space-between"
                alignItems="center"
              >
                <Text.Body>
                  <FormattedMessage {...messages.ownerMe} />
                </Text.Body>
                <PrimaryButton
                  data-testid="ownerMeButton"
                  iconLeft={<CheckBoldIcon />}
                  size="small"
                  label={this.props.intl.formatMessage(
                    messages.ownerSelectButton
                  )}
                  onClick={this.handleSelectOwnerMe}
                />
              </Spacings.Inline>
            </Card>
            <Card>
              <Spacings.Inline
                justifyContent="space-between"
                alignItems="center"
              >
                <Text.Body>
                  <FormattedMessage {...messages.ownerEmployee} />
                </Text.Body>
                <PrimaryButton
                  data-testid="ownerEmployeeButton"
                  iconLeft={<CheckBoldIcon />}
                  size="small"
                  label={this.props.intl.formatMessage(
                    messages.ownerSelectButton
                  )}
                  onClick={this.handleSelectOwnerEmployee}
                />
              </Spacings.Inline>
            </Card>
            <Card>
              <Spacings.Inline
                justifyContent="space-between"
                alignItems="center"
              >
                <Text.Body>
                  <FormattedMessage {...messages.ownerCompany} />
                </Text.Body>
                <PrimaryButton
                  data-testid="ownerCompanyButton"
                  iconLeft={<CheckBoldIcon />}
                  size="small"
                  label={this.props.intl.formatMessage(
                    messages.ownerSelectButton
                  )}
                  onClick={this.handleSelectOwnerCompany}
                />
              </Spacings.Inline>
            </Card>
          </Spacings.Inline>
        )}
        {this.props.ownerState.owner.type === OWNER_TYPES.OWNER_EMPLOYEE && (
          <Spacings.Stack scale="m">
            <Text.Headline as="h2">{this.props.title}</Text.Headline>
            <Text.Subheadline as="h5">
              {this.props.intl.formatMessage(messages.subTitle)}
            </Text.Subheadline>
            <SearchInput
              initialValue={
                this.props.employeesFetcher.searchQuery.value.searchText
              }
              onSubmit={this.handleSearchInputChange}
              placeholder={this.props.intl.formatMessage(
                messages.searchPlaceholder
              )}
            />
            {this.props.employeesFetcher.employees.results.length > 0 ? (
              <Table
                columns={createColumnsDefinition(this.props.intl)}
                rowCount={this.props.employeesFetcher.employees.results.length}
                itemRenderer={this.renderItem}
                onRowClick={(_, rowIndex) => this.handleRowClick(rowIndex)}
                shouldFillRemainingVerticalSpace={true}
                items={this.props.employeesFetcher.employees.results}
                data-testid="customer-list"
              >
                <Pagination
                  totalItems={this.props.employeesFetcher.employees.total}
                  perPage={
                    this.props.employeesFetcher.searchQuery.value.perPage
                  }
                  onPerPageChange={this.handlePerPageChange}
                  page={this.props.employeesFetcher.searchQuery.value.page}
                  onPageChange={this.handlePageChange}
                />
                <PageBottomSpacer />
              </Table>
            ) : (
              <Text.Body>
                {this.props.intl.formatMessage(messages.noSearchResults)}
              </Text.Body>
            )}
          </Spacings.Stack>
        )}
      </Fragment>
    );
  }
}

// This wrapping component is needed so we check the new results for
// redirecting automatically to th addresses step if only 1 customer is in the search
export const ConnectedOrderCreateOwnerPick = props => (
  <EmployeeDetailWrapper projectKey={props.projectKey}>
    {({ company, employee }) => (
      <B2BApolloClientContext.Consumer>
        {({ apolloClient }) => {
          return (
            <OrderCreateOwnerPickConnector
              projectKey={props.projectKey}
              storeKey={props.storeState?.value?.key}
              company={company}
              apolloClient={apolloClient}
            >
              {({ employeesFetcher }) => (
                <OrderCreateOwnerPick
                  {...props}
                  company={company}
                  employee={employee}
                  employeesFetcher={employeesFetcher}
                />
              )}
            </OrderCreateOwnerPickConnector>
          );
        }}
      </B2BApolloClientContext.Consumer>
    )}
  </EmployeeDetailWrapper>
);
ConnectedOrderCreateOwnerPick.displayName = ConnectedOrderCreateOwnerPick;
ConnectedOrderCreateOwnerPick.propTypes = {
  projectKey: PropTypes.string.isRequired,
  storeState: PropTypes.shape({
    value: PropTypes.shape({
      id: PropTypes.string.isRequired,
      key: PropTypes.string.isRequired,
      nameAllLocales: PropTypes.arrayOf(
        PropTypes.shape({
          locale: PropTypes.string.isRequired,
          value: PropTypes.string.isRequired,
        })
      ).isRequired,
    }),
    update: PropTypes.func.isRequired,
  }),
};

const mapDispatchToProps = {
  showNotification: globalActions.showNotification,
};

export default compose(
  connect(null, mapDispatchToProps),
  replaceLocationWhenDraftIsMissing,
  injectIntl
)(ConnectedOrderCreateOwnerPick);
