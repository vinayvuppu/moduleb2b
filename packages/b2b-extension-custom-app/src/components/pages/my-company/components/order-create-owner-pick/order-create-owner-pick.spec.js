import PropTypes from 'prop-types';
import React from 'react';
import { shallow } from 'enzyme';
import { Text, Card } from '@commercetools-frontend/ui-kit';
import { intlMock } from '@commercetools-local/test-utils';
import { OrderCreateOwnerPick } from './order-create-owner-pick';
import { OWNER_TYPES } from './constants';

const createConnectorProps = custom => ({
  isLoading: false,
  employees: {
    results: [],
    total: 0,
  },
  searchQuery: {
    set: jest.fn(),
    value: {
      perPage: 20,
      page: 1,
      searchText: '',
    },
  },
  ...custom,
});

const createTestProps = props => ({
  title: 'Customer',
  onActionError: jest.fn(),
  cartDraft: {},
  employeesFetcher: createConnectorProps(),
  cartDraftState: {
    value: {},
    update: jest.fn(),
  },
  storeState: {
    value: {
      id: 'store-1',
      key: 'store-1',
      nameAllLocales: [
        {
          locale: 'de',
          value: 'deutschland',
        },
      ],
    },
    update: jest.fn(),
  },
  ownerState: {
    owner: {
      type: undefined,
      company: undefined,
    },
    update: jest.fn(),
  },
  initialSelectionModalState: {
    isOpen: true,
    close: jest.fn(),
  },
  company: {
    id: 'companyId1',
    key: 'company-key-1',
    addresses: [],
  },
  employee: {
    id: 'employee-id-1',
    email: 'foo@bar.com',
    defaultBillingAddress: '1',
    defaultShippingAddress: '2',
    stores: [
      {
        key: 'store-key-1',
      },
    ],
    addresses: [
      {
        id: '1',
      },
      {
        id: '2',
      },
    ],
  },
  goToAddressSelection: jest.fn(),
  projectKey: 'test-1',
  // Connected
  showNotification: jest.fn(),
  // HoC
  intl: intlMock,

  ...props,
});

const TestRowItem = ({ children }) => <div>{children}</div>;
TestRowItem.propTypes = {
  children: PropTypes.any.isRequired,
};

describe('rendering base elements', () => {
  let props;
  let wrapper;

  describe('when no owner is selected', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<OrderCreateOwnerPick {...props} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the three option cards', () => {
      expect(wrapper.find(Card)).toHaveLength(3);
    });
  });
  describe('when owner is selected', () => {
    beforeEach(() => {
      props = createTestProps({
        ownerState: {
          owner: {
            type: OWNER_TYPES.OWNER_EMPLOYEE,
            company: undefined,
          },
          update: jest.fn(),
        },
      });
      wrapper = shallow(<OrderCreateOwnerPick {...props} />);
    });

    it('should render title', () => {
      expect(wrapper.find(Text.Headline)).toHaveProp('children', 'Customer');
    });

    it('should render SearchInput', () => {
      expect(wrapper).toRender('SearchInput');
    });

    describe('when employee search has no results', () => {
      it('should not render Table', () => {
        expect(wrapper).not.toRender('Table');
      });
      it('should render message for no results', () => {
        expect(wrapper.find('TextBody').children()).toHaveText(
          'Orders.Create.Step.Owner.noSearchResults'
        );
      });
    });

    describe('when employee search has results', () => {
      beforeEach(() => {
        props = createTestProps({
          ownerState: {
            owner: {
              type: OWNER_TYPES.OWNER_EMPLOYEE,
              company: undefined,
            },
            update: jest.fn(),
          },
          employeesFetcher: createConnectorProps({
            employees: {
              results: [
                {
                  id: '1',
                  customerNumber: 'test-1',
                  firstName: 'Jon',
                  lastName: 'Snow',
                  companyName: 'House Stark',
                  email: 'jon.snow@got.com',
                },
              ],
              total: 0,
            },
          }),
        });
        wrapper = shallow(<OrderCreateOwnerPick {...props} />);
      });

      it('should render Table', () => {
        expect(wrapper).toRender('Table');
      });

      it('should pass columns map', () => {
        expect(
          wrapper
            .find('Table')
            .prop('columns')
            .map(_ => _.key)
        ).toEqual(['employeeNumber', 'firstName', 'lastName', 'email']);
      });

      describe('should render table rows', () => {
        let instance;
        let wrapperRow;
        beforeEach(() => {
          instance = wrapper.instance();
        });

        it('should render customerNumber', () => {
          wrapperRow = shallow(
            <TestRowItem>
              {instance.renderItem({
                rowIndex: 0,
                columnKey: 'customerNumber',
              })}
            </TestRowItem>
          );
          expect(wrapperRow.text()).toBe('test-1');
        });
        it('should render firstName', () => {
          wrapperRow = shallow(
            <TestRowItem>
              {instance.renderItem({ rowIndex: 0, columnKey: 'firstName' })}
            </TestRowItem>
          );
          expect(wrapperRow.text()).toBe('Jon');
        });
        it('should render lastName', () => {
          wrapperRow = shallow(
            <TestRowItem>
              {instance.renderItem({ rowIndex: 0, columnKey: 'lastName' })}
            </TestRowItem>
          );
          expect(wrapperRow.text()).toBe('Snow');
        });
        it('should render companyName', () => {
          wrapperRow = shallow(
            <TestRowItem>
              {instance.renderItem({ rowIndex: 0, columnKey: 'companyName' })}
            </TestRowItem>
          );
          expect(wrapperRow.text()).toBe('House Stark');
        });
        it('should render email', () => {
          wrapperRow = shallow(
            <TestRowItem>
              {instance.renderItem({ rowIndex: 0, columnKey: 'email' })}
            </TestRowItem>
          );
          expect(wrapperRow.text()).toBe('jon.snow@got.com');
        });
      });
      describe('Pagination', () => {
        it('should render Pagination component', () => {
          expect(wrapper).toRender('Pagination');
        });
        it('should have `page` prop', () => {
          expect(wrapper.find('Pagination')).toHaveProp('page', 1);
        });
        it('should have `perPage` prop', () => {
          expect(wrapper.find('Pagination')).toHaveProp('perPage', 20);
        });
        it('should have `totalItems` prop', () => {
          expect(wrapper.find('Pagination')).toHaveProp(
            'totalItems',
            props.employeesFetcher.employees.total
          );
        });
      });
    });
  });
});

describe('callbacks', () => {
  let props;
  let wrapper;

  describe('when select owner type', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<OrderCreateOwnerPick {...props} />);
    });

    describe('when select OWNER_ME', () => {
      beforeEach(() => {
        wrapper.find({ 'data-testid': 'ownerMeButton' }).simulate('click');
      });

      it('should call `goToAddressSelection` prop with the employee id', () => {
        expect(props.goToAddressSelection).toHaveBeenCalledWith(
          'employee-id-1'
        );
      });

      it('should call to update the owner state', () => {
        expect(props.ownerState.update).toHaveBeenCalledWith({
          type: OWNER_TYPES.OWNER_ME,
          company: undefined,
          employeeId: 'employee-id-1',
        });
      });

      it('should call to update the cart state', () => {
        expect(props.cartDraftState.update).toHaveBeenCalledWith({
          billingAddress: { id: '1' },
          customerEmail: 'foo@bar.com',
          customerId: 'employee-id-1',
          shippingAddress: { id: '2' },
          store: {
            key: 'store-key-1',
            typeId: 'store',
          },
        });
      });
    });

    describe('when select OWNER_EMPLOYEE', () => {
      beforeEach(() => {
        wrapper
          .find({ 'data-testid': 'ownerEmployeeButton' })
          .simulate('click');
      });

      it('should call to update the owner state', () => {
        expect(props.ownerState.update).toHaveBeenCalledWith({
          type: OWNER_TYPES.OWNER_EMPLOYEE,
          company: undefined,
        });
      });
    });

    describe('when select OWNER_COMPANY', () => {
      beforeEach(() => {
        wrapper.find({ 'data-testid': 'ownerCompanyButton' }).simulate('click');
      });

      it('should call to update the owner state', () => {
        expect(props.ownerState.update).toHaveBeenCalledWith({
          type: OWNER_TYPES.OWNER_COMPANY,
          company: props.company,
        });
      });

      it('should call to cartDraftState to update the store', () => {
        expect(props.cartDraftState.update).toHaveBeenCalledWith({
          store: { key: 'store-key-1', typeId: 'store' },
          customerEmail: 'foo@bar.com',
          customerId: 'employee-id-1',
          billingAddress: {},
          shippingAddress: {},
        });
      });

      it('should call to goToAddressSelection', () => {
        expect(props.goToAddressSelection).toHaveBeenCalled();
      });
    });
  });

  describe('when user searchs for a employee', () => {
    beforeEach(() => {
      props = createTestProps({
        ownerState: {
          owner: {
            type: OWNER_TYPES.OWNER_EMPLOYEE,
            company: undefined,
          },
          update: jest.fn(),
        },
      });
      wrapper = shallow(<OrderCreateOwnerPick {...props} />);
      return wrapper.find('SearchInput').prop('onSubmit')('foo');
    });

    it('should call `set` from `searchQuery`', () => {
      expect(props.employeesFetcher.searchQuery.set).toHaveBeenCalled();
    });
  });
  describe('when user clicks in a row', () => {
    describe('when employee has addresses', () => {
      beforeEach(() => {
        props = createTestProps({
          ownerState: {
            owner: {
              type: OWNER_TYPES.OWNER_EMPLOYEE,
              company: undefined,
            },
            update: jest.fn(),
          },
          employeesFetcher: createConnectorProps({
            employees: {
              total: 1,
              results: [
                {
                  id: 'id-1',
                  email: 'foo@bar.com',
                  defaultBillingAddressId: '1',
                  defaultShippingAddressId: '2',
                  stores: [
                    {
                      key: 'store-key-1',
                    },
                  ],
                  addresses: [
                    {
                      id: '1',
                    },
                    {
                      id: '2',
                    },
                  ],
                },
              ],
            },
          }),
        });
        wrapper = shallow(<OrderCreateOwnerPick {...props} />);
        wrapper.instance().handleRowClick(0);
      });
      it('should call `update` from `cartDraftState`', () => {
        expect(props.cartDraftState.update).toHaveBeenCalledTimes(1);
      });
      it('should call goToAddressSelection', () => {
        expect(props.goToAddressSelection).toHaveBeenCalledTimes(1);
      });
    });
    describe('when employee has no addresses', () => {
      beforeEach(() => {
        props = createTestProps({
          ownerState: {
            owner: {
              type: OWNER_TYPES.OWNER_EMPLOYEE,
              company: undefined,
            },
            update: jest.fn(),
          },
          employeesFetcher: createConnectorProps({
            employees: {
              total: 1,
              results: [
                {
                  id: 'id-1',
                  email: 'foo@bar.com',
                  defaultBillingAddressId: '1',
                  defaultShippingAddressId: '2',
                  stores: [
                    {
                      key: 'store-key-1',
                    },
                  ],
                  addresses: [],
                },
              ],
            },
          }),
        });
        wrapper = shallow(<OrderCreateOwnerPick {...props} />);
        wrapper.instance().handleRowClick(0);
      });
      it('should call `update` from `cartDraftState`', () => {
        expect(props.cartDraftState.update).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('when user changes page', () => {
    beforeEach(() => {
      props = createTestProps({
        ownerState: {
          owner: {
            type: OWNER_TYPES.OWNER_EMPLOYEE,
            company: undefined,
          },
          update: jest.fn(),
        },
        employeesFetcher: createConnectorProps({
          employees: {
            results: [
              {
                id: '1',
                customerNumber: 'test-1',
                firstName: 'Jon',
                lastName: 'Snow',
                companyName: 'House Stark',
                email: 'jon.snow@got.com',
                stores: [
                  {
                    key: 'store-key-1',
                  },
                ],
              },
            ],
            total: 0,
          },
        }),
      });
      wrapper = shallow(<OrderCreateOwnerPick {...props} />);
      wrapper.find('Pagination').prop('onPageChange')(2);
    });
    it('should call `set` from `searchQuery`', () => {
      expect(props.employeesFetcher.searchQuery.set).toHaveBeenCalled();
    });
  });
  describe('when user changes items per page', () => {
    beforeEach(() => {
      props = createTestProps({
        ownerState: {
          owner: {
            type: OWNER_TYPES.OWNER_EMPLOYEE,
            company: undefined,
          },
          update: jest.fn(),
        },
        employeesFetcher: createConnectorProps({
          employees: {
            results: [
              {
                id: '1',
                customerNumber: 'test-1',
                firstName: 'Jon',
                lastName: 'Snow',
                companyName: 'House Stark',
                email: 'jon.snow@got.com',
                stores: [
                  {
                    key: 'store-key-1',
                  },
                ],
              },
            ],
            total: 0,
          },
        }),
      });
      wrapper = shallow(<OrderCreateOwnerPick {...props} />);
      wrapper.find('Pagination').prop('onPerPageChange')(50);
    });

    it('should call `set` from `searchQuery`', () => {
      expect(props.employeesFetcher.searchQuery.set).toHaveBeenCalled();
    });
  });
});
