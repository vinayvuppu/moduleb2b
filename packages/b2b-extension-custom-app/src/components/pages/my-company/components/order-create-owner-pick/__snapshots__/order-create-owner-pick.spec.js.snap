// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering base elements when no owner is selected should match snapshot 1`] = `
<Fragment>
  <Inline
    alignItems="flex-start"
    justifyContent="flex-start"
    scale="s"
  >
    <Card
      theme="light"
      type="raised"
    >
      <Inline
        alignItems="center"
        justifyContent="space-between"
        scale="s"
      >
        <TextBody>
          <FormattedMessage
            defaultMessage="Me"
            id="Orders.Create.Step.Owner.ownerMe"
            values={Object {}}
          />
        </TextBody>
        <PrimaryButton
          data-testid="ownerMeButton"
          iconLeft={<CheckBoldIcon />}
          isToggleButton={false}
          label="Orders.Create.Step.Owner.ownerSelectButton"
          onClick={[Function]}
          size="small"
          tone="primary"
          type="button"
        />
      </Inline>
    </Card>
    <Card
      theme="light"
      type="raised"
    >
      <Inline
        alignItems="center"
        justifyContent="space-between"
        scale="s"
      >
        <TextBody>
          <FormattedMessage
            defaultMessage="Company employee"
            id="Orders.Create.Step.Owner.ownerEmployee"
            values={Object {}}
          />
        </TextBody>
        <PrimaryButton
          data-testid="ownerEmployeeButton"
          iconLeft={<CheckBoldIcon />}
          isToggleButton={false}
          label="Orders.Create.Step.Owner.ownerSelectButton"
          onClick={[Function]}
          size="small"
          tone="primary"
          type="button"
        />
      </Inline>
    </Card>
    <Card
      theme="light"
      type="raised"
    >
      <Inline
        alignItems="center"
        justifyContent="space-between"
        scale="s"
      >
        <TextBody>
          <FormattedMessage
            defaultMessage="Company"
            id="Orders.Create.Step.Owner.ownerCompany"
            values={Object {}}
          />
        </TextBody>
        <PrimaryButton
          data-testid="ownerCompanyButton"
          iconLeft={<CheckBoldIcon />}
          isToggleButton={false}
          label="Orders.Create.Step.Owner.ownerSelectButton"
          onClick={[Function]}
          size="small"
          tone="primary"
          type="button"
        />
      </Inline>
    </Card>
  </Inline>
</Fragment>
`;
